### DISCUSSION TITLE
why-should-suspense-be-used

### DISCUSSION THREAD

**Discussion by giscus** [JhsK/sungkyu-v2.0#7] [General]
# why-should-suspense-be-used

사내에서 React18의 Suspense 적용하면서 얻은 인사이트와 어떠한 어려움들이 있었는지 공유합니다.

http://localhost:3000/why-should-suspense-be-used

**Comment by BO-LIKE-CHICKEN** [comment]
좋은 글 잘 읽었습니다!
Suspense와 ErrorBoundary에 대해서 잘 정리된 글을 찾다가 이런 좋은 글을 왜 못봤을까? 했는데 비교적 최근에 작성된 글 이군요 ㅎㅎ

> 2. 페이지네이션이 필요한 컴포넌트에서 Suspense를 사용시 페이지를 변경할 때마다 fallback 컴포넌트 노출 또는 깜빡임 문제 발생

저도 위와 같은 고민을 했는데 시도해보신 부분을 제외하고는 다음과 같이 접근해 보았습니다.

1. Fallback UI를 현재 리스트 크기 만큼의 더미데이터를 뿌려주는 스켈레톤 형태로 만들고 최소 딜레이를 주자.
장점: 데이터를 불러오고 있음을 유저가 인식하며, 화면이 번쩍이는 것처럼 느껴지지 않습니다.
단점:  다만, 이미 데이터를 불러왔음에도 최소한의 딜레이를 유저가 보아야 한다는 단점이 있습니다. 그리고 마지막 페이지로 이동한다고 가정한다면 어쩔 수 없이 리스트의 크기가 줄어들며 화면이 덜그럭거리게 되는 경우가 존재합니다.

2. 다음 페이지를 미리 Prefetch 하자.
장점: 화면 번쩍임이 없어질 뿐만 아니라, 유저가 빠르게 다음 페이지의 정보를 볼 수 있습니다.
단점: 확실하게 문제는 개선되지만, API 호출 수가 증가한다는 단점이 있었습니다. 만약 유저들이 1페이지에 진입해 2페이지 까지 호출하는 경우가 많다면 이 부분이 효과적일 수 있으나 그렇지 않을 경우 비용 문제를 야기합니다.
보완 방법: 따라서, 2페이지 까지 마우스 커서가 움직이거나 스크롤이 최하단에 닿았을 때 prefetch 하는 식으로 접근해보면 어떨까? 라는 생각이 듭니다.

근본적인 해결책은 아니나 저도 글을 읽고 기존의 코드를 다시 만져보면서 모든 경우에 Suspense(Query)를 통한 데이터 페칭이 올바른지에 대해서 고민이 드네요.

**Reply by JhsK** [reply]
@BO-LIKE-CHICKEN
안녕하세요! 우선 두서 없는 글 읽어주셔서 너무나 감사드립니다! 🙇‍♂️

생각해보지 못 한 다른 방법들을 소개해 주셔서 고민해보았습니다!
> 1. Fallback UI를 현재 리스트 크기 만큼의 더미데이터를 뿌려주는 스켈레톤 형태로 만들고 최소 딜레이를 주자.

최소한의 딜레이가 보이는 단점이 있지만 일관성있게 UI를 보여주고, 데이터를 불러오고 있음을 유저가 느낄 수 있다는 부분이 매력적이였습니다! 마지막 페이지로 이동했을 때 덜그럭거리는 문제만 해결할 수 있다면 좋은 방법이지 않을까 생각해요!! 저도 상황에 따라 고려해서 사용해보겠습니다!

> 2. 다음 페이지를 미리 Prefetch 하자.

Prefetch를 생각을 못 했었네요! 예전에 다른 상황에서도 Prefetch를 고려했던 적이 있었는데 말씀해주신 비용 문제가 따라오기 때문에 어느 정도까지 허용을 해서 사용해야 할 지에 대해 고민이 많아 다른 방법으로 해결했었던 기억이 있는데 말씀해주신 보완 방법을 이용해서 최소한의 비용으로 적용해보면 좋을 것 같습니다!!

기획 또는 요구사항에 따라 Suspense를 사용해야 하는 경우나 효과가 좋은 상황이라면 말씀해주신 방법을 이용해서 해결하는게 제일 효과적일 것 같습니다!! 고민했던 부분과 좋은 방법들을 소개해주셔서 너무 감사합니다!! 

저 또한 다른 좋은 방법이 생기면 공유드리겠습니다! 감사합니다!

**Comment by HanSeongLee** [comment]
잘 읽었습니다 :)

'useDefferedValue 활용'에서 깜빡이지 않으려면 useDefferedValue를 data가 아닌 queryKey에 적용해야 될 것 같습니다.
